src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,633 @@
+            pk_index,
+            timestamp: ts_end,
+        }
+    }
+
+    fn search_key(&self, key: &Self::Key) -> std::result::Result<usize, usize> {
+        let DataBatchKey {
+            pk_index,
+            timestamp,
+        } = key;
+        if *pk_index != self.pk_index() {
+            return Err(self.range().end);
+        }
+        let ts_values = timestamp_array_to_i64_slice(self.record_batch().column(1));
+        ts_values.binary_search(timestamp)
+    }
+
+    fn slice(&self, range: Range<usize>) -> Self {
+        let rb = self.rb.slice(range.start, range.len());
+        let range = 0..rb.num_rows();
+        Self {
+            pk_index: self.pk_index,
+            rb,
+            range,
+        }
+    }
+}
+
+pub struct DataNode {
+    source: DataSource,
+    current_data_batch: Option<DataBatch>,
+}
+
+impl DataNode {
+    pub(crate) fn new(source: DataSource) -> Self {
+        let current_data_batch = source.current_data_batch();
+        Self {
+            source,
+            current_data_batch: Some(current_data_batch),
+        }
+    }
+
+    fn next(&mut self) -> Result<()> {
+        let next = match &mut self.source {
+            DataSource::Buffer(b) => {
+                b.next()?;
+                if b.is_valid() {
+                    Some(b.current_data_batch())
+                } else {
+                    None
+                }
+            }
+            DataSource::Part(p) => {
+                p.next()?;
+                if p.is_valid() {
+                    Some(p.current_data_batch())
+                } else {
+                    None
+                }
+            }
+        };
+        self.current_data_batch = next;
+        Ok(())
+    }
+
+    fn current_data_batch(&self) -> &DataBatch {
+        self.current_data_batch.as_ref().unwrap()
+    }
+}
+
+pub enum DataSource {
+    Buffer(DataBufferReader),
+    Part(DataPartReader),
+}
+
+impl DataSource {
+    pub(crate) fn current_data_batch(&self) -> DataBatch {
+        match self {
+            DataSource::Buffer(buffer) => buffer.current_data_batch(),
+            DataSource::Part(p) => p.current_data_batch(),
+        }
+    }
+}
+
+impl Ord for DataNode {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let weight = self.current_data_batch().pk_index;
+        let (ts_start, sequence) = self.current_data_batch().first_row();
+        let other_weight = other.current_data_batch().pk_index;
+        let (other_ts_start, other_sequence) = other.current_data_batch().first_row();
+        (weight, ts_start, Reverse(sequence))
+            .cmp(&(other_weight, other_ts_start, Reverse(other_sequence)))
+            .reverse()
+    }
+}
+
+impl Eq for DataNode {}
+
+impl PartialEq<Self> for DataNode {
+    fn eq(&self, other: &Self) -> bool {
+        self.current_data_batch()
+            .first_row()
+            .eq(&other.current_data_batch().first_row())
+    }
+}
+
+#[allow(clippy::non_canonical_partial_ord_impl)]
```

Comment:
```suggestion
```

Reply:
> 

fixed in https://github.com/GreptimeTeam/greptimedb/pull/3361/commits/dc1c86a1ff2acf1c0cea69dc3960a661de2f7a23#diff-8e513b995e680c513011a438989b7551954f873b55a1a4ba602550e21d5d8094L340